# Maven Pizza Analysis

## Data Dictionary


### Order Details CSV

### Importing Data Sets

In [2]:
# Order Details CSV
# Importing pandas, reading csv, creating dataframe, printing last frames.
import pandas as pd
order_details = pd.read_csv('order_details.csv', index_col='order_details_id')
order_details.tail(5)

order_id       pizza_id  quantity
order_details_id                                   
48616                21348  ckn_alfredo_m         1
48617                21348  four_cheese_l         1
48618                21348   napolitana_s         1
48619                21349     mexicana_l         1
48620                21350      bbq_ckn_s         1

In [3]:
#Orders CSV
# Reading csv, creating dataframe, printing last frames.
# Date and time columns separated, pulling them together with proper format.
orders = pd.read_csv('orders.csv', index_col='order_id', parse_dates = [['date', 'time']])
orders.tail()

date_time
order_id                    
21346    2015-12-31 20:51:07
21347    2015-12-31 21:14:37
21348    2015-12-31 21:23:10
21349    2015-12-31 22:09:54
21350    2015-12-31 23:02:05

In [5]:
# Pizzas CSV
# Loading csv, creating dataframe, printing last frames.
pizzas = pd.read_csv('pizzas.csv', index_col='pizza_id')
pizzas.tail()

pizza_type_id size  price
pizza_id                               
spinach_fet_m   spinach_fet    M  16.00
spinach_fet_l   spinach_fet    L  20.25
veggie_veg_s     veggie_veg    S  12.00
veggie_veg_m     veggie_veg    M  16.00
veggie_veg_l     veggie_veg    L  20.25

In [6]:
#Pizza_type CSV
# Reading CSV, creating dataframe no index, printing
pizza_types = pd.read_csv('pizza_types.csv', index_col='pizza_type_id', encoding='unicode_escape')
pizza_types.tail()

name category  \
pizza_type_id                                               
mediterraneo             The Mediterranean Pizza   Veggie   
mexicana                      The Mexicana Pizza   Veggie   
spin_pesto               The Spinach Pesto Pizza   Veggie   
spinach_fet           The Spinach and Feta Pizza   Veggie   
veggie_veg     The Vegetables + Vegetables Pizza   Veggie   

                                                     ingredients  
pizza_type_id                                                     
mediterraneo   Spinach, Artichokes, Kalamata Olives, Sun-drie...  
mexicana       Tomatoes, Red Peppers, Jalapeno Peppers, Red O...  
spin_pesto     Spinach, Artichokes, Tomatoes, Sun-dried Tomat...  
spinach_fet    Spinach, Mushrooms, Red Onions, Feta Cheese, G...  
veggie_veg     Mushrooms, Tomatoes, Red Peppers, Green Pepper...

## Exploratory Analysis
### Order Details Dataframe

In [28]:
order_details.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48620 entries, 1 to 48620
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   order_id  48620 non-null  int64 
 1   pizza_id  48620 non-null  object
 2   quantity  48620 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 1.5+ MB


- The Order Details Dataframe has 48620 entries.
- There are no null values in this Dataframe.

In [29]:
order_details.duplicated().sum()

0

- There is no duplicate entries in this Dataframe.

In [30]:
order_details.nunique()

order_id    21350
pizza_id       91
quantity        4
dtype: int64

In [ ]:
- Too many unique entries in the object column to categorize.

#### Order Details Dataframe
|Table |Column |Data Type |Description |Key|
|-------|---------|-----------|-----------|---|
|Order Details|order_details_id|int64|Numerical designation for each unique pizza per order.|X|
|Order Details|order_id|int64|Numerical designation for each order, which may contain multiple pizzas.||
|Order Details|pizza_id|object|Snakecase combination of pizza name and size designation(S,M, or L).||
|Order Details|quantity|int64|Sum of how many of a particular pizza per order.||

### Orders Dataframe

In [7]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21350 entries, 1 to 21350
Data columns (total 1 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date_time  21350 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 333.6 KB


- The Orders Dataframe has 21350 entries.
- There are no null values in the Dataframe.

In [11]:
orders.duplicated().sum()

0

- There are no duplicated entries in this Dataframe. 

In [15]:
'''Question: How many customers do we have each day? Are there any peak hours?
- We will need to categorize the dates by days of the week. We will also need hours alone for calculation.

Question: How much money did we make this year? Can we identify any seasonality in the sales?
- We will also need to categorize the month.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.dt.date.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.dt.day_name.html#pandas.Series.dt.day_name'''

orders['month'] = orders['date_time'].dt.month_name()
orders['day'] = orders['date_time'].dt.day_name()
orders['hour'] = orders['date_time'].dt.hour
orders['week'] = orders['date_time'].dt.isocalendar().week
orders.tail()

date_time     month       day  hour  week
order_id                                                    
21346    2015-12-31 20:51:07  December  Thursday    20    53
21347    2015-12-31 21:14:37  December  Thursday    21    53
21348    2015-12-31 21:23:10  December  Thursday    21    53
21349    2015-12-31 22:09:54  December  Thursday    22    53
21350    2015-12-31 23:02:05  December  Thursday    23    53

In [19]:
orders.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21350 entries, 1 to 21350
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date_time  21350 non-null  datetime64[ns]
 1   month      21350 non-null  object        
 2   day        21350 non-null  object        
 3   hour       21350 non-null  int64         
 4   week       21350 non-null  UInt32        
dtypes: UInt32(1), datetime64[ns](1), int64(1), object(2)
memory usage: 3.2 MB


- Dataframe is utilizing 3.2 MB of memory.

In [21]:
orders.memory_usage(deep = True)

Index         170800
date_time     170800
month        1347224
day          1369889
hour          170800
week          106750
dtype: int64

- It's the "month" and "day" columns utilizing the most memory.

In [22]:
orders.nunique()

date_time    21350
month           12
day              7
hour            15
week            53
dtype: int64

- The "month" and "day" columns have few unique values. Converting to a categorical data type will improve performance.

In [25]:
#https://pandas.pydata.org/docs/reference/api/pandas.Categorical.html

orders['month'] = pd.Categorical(
    values = orders['month'], 
    categories = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
    ordered = True
)

In [26]:
orders['day'] = pd.Categorical(
    values = orders['day'],
    categories = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
    ordered = True
)

In [27]:
orders.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21350 entries, 1 to 21350
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date_time  21350 non-null  datetime64[ns]
 1   month      21350 non-null  category      
 2   day        21350 non-null  category      
 3   hour       21350 non-null  int64         
 4   week       21350 non-null  UInt32        
dtypes: UInt32(1), category(2), datetime64[ns](1), int64(1)
memory usage: 648.1 KB


- 648.1 KB is the improvement! 

#### Orders Dataframe
|Table |Column |Data Type |Description |Key|
|-------|---------|-----------|-----------|---|
|orders|order_id|int64|Numerical designation for each individual order.|X|
|orders|date_time|datetime64[ns]|The date and time an orders is placed. Prior to cooking and serving.||
|orders|month|category|Months by name, Julian calendar, January - December.||
|orders|day|category|Days of the week by name, Monday - Sunday.||
|orders|hour|int64|24-Hour Clock.||
|orders|week|UInt32|52 weeks in the year.||

In [32]:
orders.tail()

date_time     month       day  hour  week
order_id                                                    
21346    2015-12-31 20:51:07  December  Thursday    20    53
21347    2015-12-31 21:14:37  December  Thursday    21    53
21348    2015-12-31 21:23:10  December  Thursday    21    53
21349    2015-12-31 22:09:54  December  Thursday    22    53
21350    2015-12-31 23:02:05  December  Thursday    23    53

pandas.core.frame.DataFrame

Index(['order_id', 'pizza_id', 'quantity'], dtype='object')

In [5]:
# Verifying the correct data types for each column.
order_details.dtypes

order_id     int64
pizza_id    object
quantity     int64
dtype: object

In [6]:
# Checking for null values.
order_details.isnull().sum().sum()

0

### Orders CSV

In [7]:
# Reading csv, creating dataframe, printing.
orders = pd.read_csv('orders.csv', index_col=0)
print(orders)

                date      time
order_id                      
1         2015-01-01  11:38:36
2         2015-01-01  11:57:40
3         2015-01-01  12:12:28
4         2015-01-01  12:16:31
5         2015-01-01  12:21:30
...              ...       ...
21346     2015-12-31  20:51:07
21347     2015-12-31  21:14:37
21348     2015-12-31  21:23:10
21349     2015-12-31  22:09:54
21350     2015-12-31  23:02:05

[21350 rows x 2 columns]


In [8]:
# Verfiying dataframe.
type(orders)

pandas.core.frame.DataFrame

In [9]:
# Checking column datatypes.
orders.dtypes

date    object
time    object
dtype: object

##### Date and Time columns are in the incorrect formats.

In [10]:
# Converting date and time columns to datetime data types.
orders['date'] = pd.to_datetime(orders['date'])
orders['time'] = pd.to_datetime(orders['time']).dt.time
orders.dtypes

date    datetime64[ns]
time            object
dtype: object

In [11]:
print(orders.head(5))

               date      time
order_id                     
1        2015-01-01  11:38:36
2        2015-01-01  11:57:40
3        2015-01-01  12:12:28
4        2015-01-01  12:16:31
5        2015-01-01  12:21:30


In [12]:
# Checking for null values.
orders.isnull().sum().sum()

0

### Pizzas CSV

In [13]:
# Loading csv, no index, creating dataframe, printing top and bottom.
pizzas = pd.read_csv('pizzas.csv', index_col=None)
print(pizzas)

         pizza_id pizza_type_id size  price
0       bbq_ckn_s       bbq_ckn    S  12.75
1       bbq_ckn_m       bbq_ckn    M  16.75
2       bbq_ckn_l       bbq_ckn    L  20.75
3      cali_ckn_s      cali_ckn    S  12.75
4      cali_ckn_m      cali_ckn    M  16.75
..            ...           ...  ...    ...
91  spinach_fet_m   spinach_fet    M  16.00
92  spinach_fet_l   spinach_fet    L  20.25
93   veggie_veg_s    veggie_veg    S  12.00
94   veggie_veg_m    veggie_veg    M  16.00
95   veggie_veg_l    veggie_veg    L  20.25

[96 rows x 4 columns]


In [14]:
#Verifying dataframe
type(pizzas)

pandas.core.frame.DataFrame

In [15]:
#Verifying column data types
pizzas.dtypes

pizza_id          object
pizza_type_id     object
size              object
price            float64
dtype: object

In [16]:
# Checking for na values
pizzas.isnull().sum().sum()

0

### Pizza Type CSV

In [17]:
# Reading CSV, creating dataframe no index, printing
pizza_types = pd.read_csv('pizza_types.csv', index_col=0, encoding='unicode_escape')
print(pizza_types.head(5))

                                       name category  \
pizza_type_id                                          
bbq_ckn          The Barbecue Chicken Pizza  Chicken   
cali_ckn       The California Chicken Pizza  Chicken   
ckn_alfredo       The Chicken Alfredo Pizza  Chicken   
ckn_pesto           The Chicken Pesto Pizza  Chicken   
southw_ckn      The Southwest Chicken Pizza  Chicken   

                                                     ingredients  
pizza_type_id                                                     
bbq_ckn        Barbecued Chicken, Red Peppers, Green Peppers,...  
cali_ckn       Chicken, Artichoke, Spinach, Garlic, Jalapeno ...  
ckn_alfredo    Chicken, Red Onions, Red Peppers, Mushrooms, A...  
ckn_pesto      Chicken, Tomatoes, Red Peppers, Spinach, Garli...  
southw_ckn     Chicken, Tomatoes, Red Peppers, Red Onions, Ja...  


In [18]:
#Verifying Dataframe
type(pizza_types)

pandas.core.frame.DataFrame

In [19]:
#Ensuring all the data types are strings(objects)
pizza_types.dtypes

name           object
category       object
ingredients    object
dtype: object

In [20]:
#Making sure there's no null values.
pizza_types.isnull().sum().sum()

0

In [21]:
#Splitting the ingredients column into separate columns to aid in analysis
pizza_types[['ing_1', 'ing_2', 'ing_3', 'ing_4', 'ing_5', 'ing_6', 'ing_7', 'ing_8']] = pizza_types['ingredients'].str.split(',', expand=True)
print(pizza_types.head(5))

                                       name category  \
pizza_type_id                                          
bbq_ckn          The Barbecue Chicken Pizza  Chicken   
cali_ckn       The California Chicken Pizza  Chicken   
ckn_alfredo       The Chicken Alfredo Pizza  Chicken   
ckn_pesto           The Chicken Pesto Pizza  Chicken   
southw_ckn      The Southwest Chicken Pizza  Chicken   

                                                     ingredients  \
pizza_type_id                                                      
bbq_ckn        Barbecued Chicken, Red Peppers, Green Peppers,...   
cali_ckn       Chicken, Artichoke, Spinach, Garlic, Jalapeno ...   
ckn_alfredo    Chicken, Red Onions, Red Peppers, Mushrooms, A...   
ckn_pesto      Chicken, Tomatoes, Red Peppers, Spinach, Garli...   
southw_ckn     Chicken, Tomatoes, Red Peppers, Red Onions, Ja...   

                           ing_1         ing_2           ing_3        ing_4  \
pizza_type_id                                      

In [22]:
#Removing the ingredients column now that I have separated ingredient columns.
pizza_types_exp = pizza_types.drop(['ingredients'], axis=1)
print(pizza_types_exp.head(5))
print(pizza_types_exp.tail(5))

                                       name category              ing_1  \
pizza_type_id                                                             
bbq_ckn          The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
cali_ckn       The California Chicken Pizza  Chicken            Chicken   
ckn_alfredo       The Chicken Alfredo Pizza  Chicken            Chicken   
ckn_pesto           The Chicken Pesto Pizza  Chicken            Chicken   
southw_ckn      The Southwest Chicken Pizza  Chicken            Chicken   

                      ing_2           ing_3        ing_4              ing_5  \
pizza_type_id                                                                 
bbq_ckn         Red Peppers   Green Peppers     Tomatoes         Red Onions   
cali_ckn          Artichoke         Spinach       Garlic   Jalapeno Peppers   
ckn_alfredo      Red Onions     Red Peppers    Mushrooms      Asiago Cheese   
ckn_pesto          Tomatoes     Red Peppers      Spinach             Garlic   


### Joining Dataframes

In [23]:
#Joining the orders and order_details dataframes.
orders_comb_df = pd.merge(order_details, orders, on='order_id')
print(orders_comb_df)

       order_id       pizza_id  quantity       date      time
0             1     hawaiian_m         1 2015-01-01  11:38:36
1             2  classic_dlx_m         1 2015-01-01  11:57:40
2             2  five_cheese_l         1 2015-01-01  11:57:40
3             2    ital_supr_l         1 2015-01-01  11:57:40
4             2     mexicana_m         1 2015-01-01  11:57:40
...         ...            ...       ...        ...       ...
48615     21348  ckn_alfredo_m         1 2015-12-31  21:23:10
48616     21348  four_cheese_l         1 2015-12-31  21:23:10
48617     21348   napolitana_s         1 2015-12-31  21:23:10
48618     21349     mexicana_l         1 2015-12-31  22:09:54
48619     21350      bbq_ckn_s         1 2015-12-31  23:02:05

[48620 rows x 5 columns]


In [24]:
#Joining the pizzas and pizza types Dataframes.
pizzas_comb_df = pd.merge(pizzas, pizza_types_exp, on='pizza_type_id')
print(pizzas_comb_df)

         pizza_id pizza_type_id size  price  \
0       bbq_ckn_s       bbq_ckn    S  12.75   
1       bbq_ckn_m       bbq_ckn    M  16.75   
2       bbq_ckn_l       bbq_ckn    L  20.75   
3      cali_ckn_s      cali_ckn    S  12.75   
4      cali_ckn_m      cali_ckn    M  16.75   
..            ...           ...  ...    ...   
91  spinach_fet_m   spinach_fet    M  16.00   
92  spinach_fet_l   spinach_fet    L  20.25   
93   veggie_veg_s    veggie_veg    S  12.00   
94   veggie_veg_m    veggie_veg    M  16.00   
95   veggie_veg_l    veggie_veg    L  20.25   

                                 name category              ing_1  \
0          The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
1          The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
2          The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
3        The California Chicken Pizza  Chicken            Chicken   
4        The California Chicken Pizza  Chicken            Chicken   
..                   

In [25]:
pizzas_comb_df.dtypes

pizza_id          object
pizza_type_id     object
size              object
price            float64
name              object
category          object
ing_1             object
ing_2             object
ing_3             object
ing_4             object
ing_5             object
ing_6             object
ing_7             object
ing_8             object
dtype: object

In [27]:
#Putting it into one dataframe for Tableau Public.
combined_pizza_order_df = pd.merge(pizzas_comb_df, orders_comb_df, on='pizza_id')
print(combined_pizza_order_df)

           pizza_id pizza_type_id size  price  \
0         bbq_ckn_s       bbq_ckn    S  12.75   
1         bbq_ckn_s       bbq_ckn    S  12.75   
2         bbq_ckn_s       bbq_ckn    S  12.75   
3         bbq_ckn_s       bbq_ckn    S  12.75   
4         bbq_ckn_s       bbq_ckn    S  12.75   
...             ...           ...  ...    ...   
48615  veggie_veg_l    veggie_veg    L  20.25   
48616  veggie_veg_l    veggie_veg    L  20.25   
48617  veggie_veg_l    veggie_veg    L  20.25   
48618  veggie_veg_l    veggie_veg    L  20.25   
48619  veggie_veg_l    veggie_veg    L  20.25   

                                    name category              ing_1  \
0             The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
1             The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
2             The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
3             The Barbecue Chicken Pizza  Chicken  Barbecued Chicken   
4             The Barbecue Chicken Pizza  Chicken  

In [31]:
combined_pizza_order_df = combined_pizza_order_df[['date', 'time', 'order_id', 'quantity', 'name', 'pizza_type_id', 'size', 'price', 'category', 'ing_1', 'ing_2', 'ing_3', 'ing_4', 'ing_5', 'ing_6', 'ing_7', 'ing_8']]
combined_pizza_order_df.dtypes

date             datetime64[ns]
time                     object
order_id                  int64
quantity                  int64
name                     object
pizza_type_id            object
size                     object
price                   float64
category                 object
ing_1                    object
ing_2                    object
ing_3                    object
ing_4                    object
ing_5                    object
ing_6                    object
ing_7                    object
ing_8                    object
dtype: object

In [33]:
combined_pizza_order_df.to_csv('combined_pizza_order.csv', index=False)